sdk/servicebus/Azure.Messaging.ServiceBus/src/Sender/ServiceBusMessageBatch.cs

```diff
@@ -91,7 +96,9 @@ public bool TryAddMessage(ServiceBusMessage message)
             lock (_syncGuard)
             {
                 AssertNotLocked();
-                return InnerBatch.TryAddMessage(message);
+
+                _scopeFactory.InstrumentMessage(message);
```

Comment:
I didn't see any issues with just instrumenting the message here as opposed to waiting for the send. In either case we mutate the message itself (not just the underlying AMQP message), so there doesn't seem to be a downside to potentially doing it a bit earlier.

Reply:
With this approach, the batch can be manipulated and the size exceeded.  

Event Hubs clones the message as a defensive measure to prevent changes after the measurement was done.   The clone, not the original event, is stored in the batch because the caller holds a reference to the original and can modify it which invalidates the measurement.    If the caller adds a message using `TryAdd` and then adds several user properties to the message, the size of the batch is now invalid and larger than was calculated.

Reply:
Ideally, we would be serializing to AMQP here and not cloning the event/message model, but we ran into issues with the AMQP library when retrying using that approach.    #9569 tracks revisiting this and, hopefully, avoiding the need for paying the serialization cost twice.

Reply:
So my understanding is that we either risk letting the batch get too large, or we ignore the changes a user makes to a message after they have added to the batch. I would prefer the former because if the batch gets too large, the user will be made aware of it via exception, whereas in the latter, they may not realize that the data they are setting is being ignored.

Reply:
I disagree and would recommend the first.   My thoughts:

- Cloning keeps the behavior consistent with the Service Bus behavior in _most_ of the other languages.  JS and Python store the AMQP message.  I think Java holds a reference to the message if I'm reading correctly. 

- Cloning keeps the behavior consistent with T1 and T2 Event Hubs in .NET.

- Cloning allows us to performance tune by caching the AMQP message instead of the model type, reducing the cost to a single serialization instead of paying it twice.

- Not cloning violates the guarantee made by the batch that `TryAdd` offers a safe operation; if it says `true` then you won't have an exception later.

- Not cloning leads to potential data corruption if the message is added to two different batches at the same time, as the instrumentation will fight for the same property slot.

Reply:
The risk of data loss by cloning seems like a bigger issue than cross language or cross service consistency.

Reply:
> Not cloning leads to potential data corruption if the message is added to two different batches at the same time, as the instrumentation will fight for the same property slot.

Wouldn't this be the case if you do a simultaneous sends with the same message? This doesn't seem like a valid scenario.

Reply:
Discussed offline and came to the conclusion that either behavior has its tradeoffs. I prefer leaving it as is especially since this has already GA'ed.